pkg/virt-launcher/virtwrap/network/podinterface.go

```diff
@@ -1136,19 +1154,21 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
 			strings.ToLower(port.Protocol),
 			"dport",
 			strconv.Itoa(int(port.Port)),
-			Handler.GetNFTIPString(proto), "saddr", getLoopbackAdrress(proto),
+			Handler.GetNFTIPString(proto), "daddr", b.getVifIpByProtocol(proto),
```

Comment:
> Doesn't it mean that even if there is no istio at all, the source address of the packets arrive to the vm will be the ip of the bridge (10.0.2.1)?
I think it would be better to keep the original sending ip when possible.

Yes, you're right, the consequence of this is that the VM would see the origin as 10.0.2.1.

> BTW, how the opposite direction works? If the vm sends a reply who puts the original sender ip address on the packet? Envoy?

IIUC it's handled by the SNAT rule.
http://alumni.cse.ucsc.edu/~jed/articles/iptables/iptables-tutorial.html#RCTEST-IPTABLESTXT section 3.3 is the best explanation I could find, but let me try to elaborate.

So, the nat table is traversed only for the first packet of a new connection. As this table is traversed, an underlying nf_conntrack table is constructed. This table contains active connections (connection is there identified by: protocol, src/dst IP and src/dst port).
You can examine these active connections by looking at /proc/net/nf_conntrack
It contains something like this:
```
ipv4     2 tcp      6 119 TIME_WAIT src=192.168.66.102 dst=10.244.140.83 sport=34374 dport=8080 src=10.0.2.2 dst=192.168.66.102 sport=8080 dport=34374 [ASSURED] mark=0 secctx=system_u:object_r:unlabeled_t:s0 zone=0 use=2
ipv4     2 tcp      6 114 TIME_WAIT src=192.168.66.102 dst=10.244.140.83 sport=34354 dport=8080 src=10.0.2.2 dst=192.168.66.102 sport=8080 dport=34354 [ASSURED] mark=0 secctx=system_u:object_r:unlabeled_t:s0 zone=0 use=2
ipv4     2 tcp      6 99 TIME_WAIT src=192.168.66.102 dst=10.244.140.83 sport=34252 dport=8080 src=10.0.2.2 dst=192.168.66.102 sport=8080 dport=34252 [ASSURED] mark=0 secctx=system_u:object_r:unlabeled_t:s0 zone=0 use=2
```

These are just 3 curl request to a hello world nc server running in the VM issued from the node.
You can see that at the end of line, there are 2 uses of each connection - request + response.
The response is therefore not going through the nat table at all, it's source IP is determined by the SNAT table mapping from the request destination IP and port.